src/basic/string-util.c

```diff
@@ -1023,6 +1023,34 @@ int free_and_strdup(char **p, const char *s) {
         return 1;
 }
 
+int free_and_strndup(char **p, const char *s, size_t l) {
+        char *t;
+
+        assert(p);
+        assert(s || l == 0);
+
+        /* Replaces a string pointer with an strndup()ed new string,
+         * freeing the old one. */
+
+        if (!*p && !s)
+                return 0;
+
+        if (!!*p == !!s && strneq(*p, s, l) && (l > strlen(*p) || !(*p)[l]))
```

Comment:
the first part of this test is unnecessarily complex and misleading I think. I mean, one of *p or s must be non-NULL anyway, due to the if check a few lines up, hence  the only way how !!*p  == !!s can hold is the "*p && s" case and I think writing that would make things a lot simpler and easier to grok.

also the strlen() part is flawed too. Let's say *p is "foo" and s is "foobar" and l is 6, then  this would shortcut things and it really should not. 

Also, checking for NUL termination with `!(*p)[l]` is not pretty, I'd do an != 0 check here...

Reply:
> s the "*p && s" case and I think writing that would make things a lot simpler and easier to grok.

I added the check above later and didn't think to change this part.

> also the strlen() part is flawed too. Let's say *p is "foo" and s is "foobar" and l is 6, then this would shortcut things and it really should not.

The `strneq` check is earlier and it already fails, so we don't get to strlen. I added this case to the tests and is passes. PTAL.